File: worker/query_wrapper.py
import logging
from typing import TypeVar, Callable, Any, Optional

logger = logging.getLogger(__name__)

T = TypeVar('T')

def safe_query(
    query_fn: Callable[..., T],
    endpoint: str = "unknown",
    method: str = "UNKNOWN",
    error_code: str = "E_QUERY_FAILED",
    *args: Any,
    **kwargs: Any
) -> T:
    """
    Wraps any query or external RPC execution to automatically catch,
    log, and manage errors following the `spec/03-error-manage` guidelines.
    Uses the standard `logging` library to explicitly log failures.
    """
    try:
        return query_fn(*args, **kwargs)
    except Exception as e:
        logger.error(
            "Query failed: %s", str(e),
            exc_info=True,
            extra={
                "error_code": error_code,
                "endpoint": endpoint,
                "method": method,
                "source": "safe_query"
            }
        )
        # Re-raise the exception after logging it so the caller knows it failed
        raise e

File: src/lib/utils/query-wrapper.ts
import { ClientLogger } from "@/lib/observability/client-logger";
import { useErrorStore } from "@/lib/errors/errorStore";

/**
 * Wraps any promise-based query (fetch, axios, RPC) to automatically catch
 * and log errors using the guidelines in `spec/03-error-manage`.
 * Connects to `useErrorStore` for unified error handling.
 */
export async function QueryWrapper<T>(
  queryFn: () => Promise<T>,
  context: {
    endpoint?: string;
    method?: string;
    correlationId?: string;
    suppressCapture?: boolean;
    errorCode?: string;
  } = {},
): Promise<T | null> {
  try {
    const result = await queryFn();

    return result;
  } catch (error) {
    if (!context.suppressCapture) {
      try {
        useErrorStore.getState().captureError(
          error,
          {
            endpoint: context.endpoint || "unknown",
            method: context.method || "UNKNOWN",
            correlationId: context.correlationId,
            source: "QueryWrapper",
          },
          context.errorCode || "E_QUERY_FAILED",
        );
      } catch (storeError) {
        ClientLogger.error("[QueryWrapper] Failed to push to errorStore", storeError);
      }
    }
    ClientLogger.error("[QueryWrapper] Query failed:", error, context);

    throw error;
  }
}
